### PR TITLE
Add a draft implementation for labelled adjacency maps

### PR DESCRIPTION
**Pull request by o1lo01ol1o**
@snowleopard See what you think of this.  I'm getting a hang on the `edgeSet` test, not sure why yet.

### CODE REVIEW

**Comment by snowleopard** [comment]
@o1lo01ol1o I only had time for a brief look today, but it looks great! 

To make the PR more self-contained, I decided to merge the `edge-labels` branch in #114, so you can focus only on `LabelledAdjacencyMap` in this PR.

**Comment by o1lo01ol1o** [comment]
Sure, cool.  

There's a couple of outstanding `toGraph` instance methods (ie, `toIntMap` etc don't really makes sense), there aren't any tests to explicitly test the semantics of the labels and the `show` instance should probably make use of a `Dioid e` `show` somehow, but  I thought I'd better wait for you're direction on how to address these issues.  (I'm reasonably sure that I didn't change any of the asymptotics on time / space complexity but I could have made an incorrect assumption about how maps and sets are fused. )  Let me know how to proceed.

**Comment by nobrakal** [comment]
@o1lo01ol1o I think `test/Algebra/Graph/Test/.DS_Store` was added too quickly ^^.

Great work !

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 1]
```diff
@@ -0,0 +1,780 @@
...
+-- Complexity: /O(n)/ time.
+--
+-- @
+-- edgeCount 'empty'      == 0
+-- edgeCount ('vertex' x) == 0
+-- edgeCount ('edge' x y) == 1
+-- edgeCount            == 'length' . 'edgeList'
+-- @
+edgeCount :: LabelledAdjacencyMap a e -> Int
+edgeCount = Map.foldr (\es r -> (Map.size es + r)) 0 . labelledAdjacencyMap
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: Note that Alga HEAD has a different implementation of this function for `AdjacencyMap`: 

```
edgeCount = getSum . foldMap (Sum . Set.size) . adjacencyMap
```

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+-- Complexity: /O((n + m) * log(m))/ time and /O(m)/ memory.
+--
+-- @
+-- edgeSet 'empty'      == Set.'Set.empty'
+-- edgeSet ('vertex' x) == Set.'Set.empty'
+-- edgeSet ('edge' x y) == Set.'Set.singleton' (x,y)
+-- edgeSet . 'edges'    == Set.'Set.fromList'
+-- @
+edgeSet :: Ord a => LabelledAdjacencyMap a e -> Set (a, a)
+edgeSet =
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: This could be simply `edgeSet = Set.fromAscList . edgeList`, as in `AdjacencyMap`. Or is this implementation faster?

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 3]
```diff
@@ -0,0 +1,780 @@
...
+-- given list.
+--
+-- @
+-- starTranspose x []    == 'vertex' x
+-- starTranspose x [y]   == 'edge' y x
+-- starTranspose x [y,z] == 'edges' [(y,x), (z,x)]
+-- starTranspose x ys    == 'connect' ('vertices' ys) ('vertex' x)
+-- starTranspose x ys    == 'transpose' ('star' x ys)
+-- @
+starTranspose :: (Ord a, Dioid e) => a -> [a] -> LabelledAdjacencyMap a e
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: Note that we are likely to remove `starTranspose` from the API, see #107.

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 4]
```diff
@@ -0,0 +1,780 @@
...
+-- replaceVertex x y            == 'mergeVertices' (== x) y
+-- @
+replaceVertex
+        :: Ord a
+        => a
+        -> a
+        -> LabelledAdjacencyMap a e
+        -> LabelledAdjacencyMap a e
+replaceVertex u v = gmap $ \w -> if w == u then v else w
+-- TODO: replaceVertexWithLabel?
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: Not sure what would `replaceVertexWithLabel` do?

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 5]
```diff
@@ -0,0 +1,780 @@
...
+-- TODO: replaceLabel
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: Yes, `replaceLabel` looks like a useful function.

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 6]
```diff
@@ -0,0 +1,780 @@
...
+-- mergeVertices odd  1 (3 + 4 * 5) == 4 * 1
+-- @
+mergeVertices
+        :: Ord a
+        => (a -> Bool)
+        -> a
+        -> LabelledAdjacencyMap a e
+        -> LabelledAdjacencyMap a e
+mergeVertices p v = gmap $ \u -> if p u then v else u
+-- TODO: mergeVerticesAndLabel p v l =
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: What would `mergeVerticesAndLabel` do?

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 7]
```diff
@@ -0,0 +1,780 @@
...
+transpose (LAM m) = LAM $ Map.foldrWithKey combine vs m
+    where
+        combine
+                :: (Ord a, Dioid e)
+                => a
+                -> Map.Map a e
+                -> Map.Map a (Map.Map a e)
+                -> Map.Map a (Map.Map a e)
+        combine v es = Map.unionWith
+                (Map.unionWith (/\))
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: Should this be `\/`?

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 8]
```diff
@@ -0,0 +1,780 @@
...
+--
+-- @
+-- gemap f 'empty'      == 'empty'
+-- gemap f ('vertex' x) == 'vertex' (f x)
+-- gemap f ('edge' x y) == 'edge' (f x) (f y)
+-- gemap id           == id
+-- gemap f . gmap g   == gmap (f . g)
+-- @
+gemap :: (e -> e') -> LabelledAdjacencyMap a e -> LabelledAdjacencyMap a e'
+gemap f = LAM . Map.map (Map.map f) . labelledAdjacencyMap
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: Let's use `fgl`'s naming here -- `emap`?

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 9]
```diff
@@ -0,0 +1,780 @@
...
+-- @
+topSort :: Ord a => LabelledAdjacencyMap a e -> Maybe [a]
+topSort m = if isTopSortOf result m then Just result else Nothing
+  where
+    result = Typed.topSort (Typed.fromLabelledAdjacencyMap m)
+
+-- | Check if a given list of vertices is a valid /topological sort/ of a graph.
+--
+-- @
+-- isTopSort [3, 1, 2] (1 * 2 + 3 * 1) == True
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: Typos: `isTopSort` -> `isTopSortOf`.

File: src/Algebra/Graph/LabelledAdjacencyMap/Internal.hs [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    -- | The /adjacency map/ of the graph: each vertex is associated with a set
+    -- of its direct successors.
+    labelledAdjacencyMap :: Map a (Map a e) } deriving Eq
+
+lAdjacencyMap :: LabelledAdjacencyMap a e -> Map a (Map a e)
+lAdjacencyMap = labelledAdjacencyMap
+
+
+
+instance (Ord a, Show a, Ord e, Show e, Dioid e) => Show (LabelledAdjacencyMap a e) where
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap/Internal.hs]: We need to figure out a way to visualise labelled graphs with edge labels.

File: src/Algebra/Graph/LabelledAdjacencyMap/Internal.hs [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+instance (Ord a, Num a, Dioid e) => Num (LabelledAdjacencyMap a e) where
+  fromInteger = vertex . fromInteger
+  (+) = overlay
+  (*) = connect
+  signum = const empty
+  abs = id
+  negate = id
+
+instance (NFData a, NFData e) => NFData (LabelledAdjacencyMap a e) where
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap/Internal.hs]: Can this instance be derived?

File: src/Algebra/Graph/ToGraph.hs [hunk 12]
```diff
@@ -308,6 +328,14 @@ class ToGraph t where
     toAdjacencyMap :: Ord (ToVertex t) => t -> AM.AdjacencyMap (ToVertex t)
     toAdjacencyMap = foldg AM.empty AM.vertex AM.overlay AM.connect
 
+    -- | Convert a value to the corresponding 'LAM.LabelledAdjacencyMap'.
+    --
+    -- @
+    -- toAdjacencyMap == 'foldg' 'LAM.empty' 'LAM.vertex' 'LAM.overlay' 'LAM.connect'
+    -- @
+    toLabelledAdjacencyMap :: (Ord (ToVertex t), Dioid e )=> t -> LAM.LabelledAdjacencyMap (ToVertex t) e
+    toLabelledAdjacencyMap = foldg LAM.empty LAM.vertex LAM.overlay LAM.connect
```
Comment (snowleopard) [inline src/Algebra/Graph/ToGraph.hs]: I have a feeling we'll actually need to add `class ToGraph t => ToLabelledGraph t` type class for this, with an additional associated type `ToLabel`. Let's also do this in a separate PR.

File: src/Data/Graph/Typed.hs [hunk 13]
```diff
@@ -79,6 +81,29 @@ fromAdjacencyIntMap (AIM.AM m) = GraphKL
...
+-- If @fromLabelledAdjacencyMap g == h@ then the following holds:
+--
+-- @
+-- map ('fromVertexKL' h) ('Data.Graph.vertices' $ 'toGraphKL' h)                               == 'Algebra.Graph.AdjacencyMap.vertexList' g
+-- map (\\(x, y) -> ('fromVertexKL' h x, 'fromVertexKL' h y)) ('Data.Graph.edges' $ 'toGraphKL' h) == 'Algebra.Graph.fromLabelledAdjacencyMap.edgeList' g
+--
+-- 'toGraphKL' (fromLabelledAdjacencyMap (1 * 2 + 3 * 1)) == 'array' (0,2) [(0,[1]),(1,[]),(2,[0])]
+-- 'toGraphKL' (fromLabelledAdjacencyMap (1 * 2 + 2 * 1)) == 'array' (0,1) [(0,[1]),(1,[0])]
+-- @
+fromLabelledAdjacencyMap :: Ord a => LAM.LabelledAdjacencyMap a e -> GraphKL a
```
Comment (snowleopard) [inline src/Data/Graph/Typed.hs]: I think we should instead add a conversion `LabelledAdjacencyMap a e -> AdjacencyMap a` to the `LabelledAdjacencyMap` module and then there is no need to extend this one.

**Comment by snowleopard** [comment]
@o1lo01ol1o Great work! I've left a bunch of comments above (nothing major).

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 14]
```diff
@@ -0,0 +1,780 @@
...
+-- dfsForest ('vertices' vs)               == map (\\v -> Node v []) ('Data.List.nub' $ 'Data.List.sort' vs)
+-- 'dfsForestFrom' ('vertexList' x) x        == dfsForest x
+-- dfsForest $ 3 * (1 + 4) * (1 + 5)     == [ Node { rootLabel = 1
+--                                                 , subForest = [ Node { rootLabel = 5
+--                                                                      , subForest = [] }]}
+--                                          , Node { rootLabel = 3
+--                                                 , subForest = [ Node { rootLabel = 4
+--                                                                      , subForest = [] }]}]
+-- @
+dfsForest :: (Ord a) => LabelledAdjacencyMap a e -> Forest a
```
Comment (snowleopard) [inline src/Algebra/Graph/LabelledAdjacencyMap.hs]: If you make `LabelledAdjacencyMap` an instance of `ToGraph` then there is no need to reimplement all this -- just use the default `ToGraph` implementations.

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 15]
```diff
@@ -0,0 +1,780 @@
...
+-- overlays [x]       == x
+-- overlays [x,y]     == 'overlay' x y
+-- overlays           == 'foldr' 'overlay' 'empty'
+-- 'isEmpty' . overlays == 'all' 'isEmpty'
+-- @
+overlays
+        :: (Ord a, Dioid e)
+        => [LabelledAdjacencyMap a e]
+        -> LabelledAdjacencyMap a e
+overlays = LAM . Map.unionsWith (Map.unionWith (/\)) . map labelledAdjacencyMap
```
[the remark this answers is not in the document]
Comment (o1lo01ol1o) [reply src/Algebra/Graph/LabelledAdjacencyMap.hs]: Looks like my search-replace got some dyslexia.  Will do.

File: src/Algebra/Graph/LabelledAdjacencyMap.hs [hunk 16]
```diff
@@ -0,0 +1,780 @@
...
+-- in Haskell. See <https://github.com/snowleopard/alga-paper this paper> for the
+-- motivation behind the library, the underlying theory, and implementation details.
+--
+-- This module defines the 'LabelledAdjacencyMap' data type, as well as associated
+-- operations and algorithms. 'LabelledAdjacencyMap' is an instance of the 'C.Graph' type
+-- class, which can be used for polymorphic graph construction and manipulation.
+-- "Algebra.Graph.IntAdjacencyMap" defines adjacency maps specialised to graphs
+-- with @Int@ vertices.
+-----------------------------------------------------------------------------
+module Algebra.Graph.LabelledAdjacencyMap
```
[the remark this answers is not in the document]
Comment (o1lo01ol1o) [reply src/Algebra/Graph/LabelledAdjacencyMap.hs]: 👍

File: src/Algebra/Graph/LabelledAdjacencyMap.hs (hunk 2 above)
Comment (o1lo01ol1o) [reply src/Algebra/Graph/LabelledAdjacencyMap.hs]: I doubt it's faster, it was based on one of the previous implementations in the `AdjacncyMap` implementation.

File: src/Algebra/Graph/LabelledAdjacencyMap.hs (hunk 4 above)
Comment (o1lo01ol1o) [reply src/Algebra/Graph/LabelledAdjacencyMap.hs]: These are probably not general enough warrant the addition.   My thought was that if you replace a vertex, you may want to do something specific to its labels as well.

File: src/Algebra/Graph/LabelledAdjacencyMap.hs (hunk 8 above)
Comment (o1lo01ol1o) [reply src/Algebra/Graph/LabelledAdjacencyMap.hs]: kk

File: src/Algebra/Graph/LabelledAdjacencyMap.hs (hunk 9 above)
Comment (o1lo01ol1o) [reply src/Algebra/Graph/LabelledAdjacencyMap.hs]: 👍

File: src/Algebra/Graph/LabelledAdjacencyMap/Internal.hs (hunk 10 above)
Comment (o1lo01ol1o) [reply src/Algebra/Graph/LabelledAdjacencyMap/Internal.hs]: Yea, I'll try to come up with a way of doing this when I get a second.

**Comment by snowleopard** [comment]
@o1lo01ol1o Hi there! Do you think you'll have time to finish off this PR this week? I'm currently working on edge-labelled graphs and starting to need some of your code :) If you are too busy to address all my comments, I am happy to merge a partially complete implementation with a bunch of `TODO`s left and then address them myself. Let me know what you prefer.

**Comment by snowleopard** [comment]
Thanks @o1lo01ol1o! I will probably do some changes directly in this PR in the next couple of days and will then merge.

**Comment by snowleopard** [comment]
@o1lo01ol1o Hi there, as you can see I merged this PR after some substantial simplifications: 

* Functions like `path` do not seem to be useful for labelled graphs. Instead, I plan to add a function for converting any unlabelled graph into the corresponding labelled version, by specifying the label to use for every edge. In this way we can reuse unlabelled versions of functions like `path`.

* I removed most of the tests to make this PR a bit more manageable.

* I decided to have a slightly different API for unlabelled graphs, for example, functions `connect` and `edge` have an additional argument for the label. I think in practice `connect one` and `edge one` will be used very rarely, so they probably don't deserve a separate function.

Thanks again for your initial work on this PR! I hope the result meets your initial requirements. If not, don't hesitate to discuss possible changes in #109, which I'll keep open until documentation and tests are finished.
